gsd/gsd.c

```diff
@@ -749,51 +679,872 @@ int gsd_read_header(struct gsd_handle* handle)
+
+    @param id [out] ID of the new name
+    @param handle handle to the open gsd file
+    @param name string name
+
+    Append a name to the names in the current frame. gsd_end_frame() will add this list to the
+    file names.
+
+    @return
+      - GSD_SUCCESS (0) on success. Negative value on failure:
+      - GSD_ERROR_IO: IO error (check errno).
+      - GSD_ERROR_MEMORY_ALLOCATION_FAILED: Unable to allocate memory.
+      - GSD_ERROR_FILE_MUST_BE_WRITABLE: File must not be read only.
+*/
+inline static int gsd_append_name(uint16_t* id, struct gsd_handle* handle, const char* name)
+{
+    if (handle->open_flags == GSD_OPEN_READONLY)
+    {
+        return GSD_ERROR_FILE_MUST_BE_WRITABLE;
+    }
+
+    if (handle->file_names.n_names + handle->frame_names.n_names == UINT16_MAX)
+    {
+        // no more names may be added
+        return GSD_ERROR_NAMELIST_FULL;
+    }
+
+    // Provide the ID of the new name
+    *id = (uint16_t)(handle->file_names.n_names + handle->frame_names.n_names);
+
+    if (handle->header.gsd_version < gsd_make_version(2, 0))
+    {
+        // v1 files always allocate GSD_NAME_SIZE bytes for each name and put a NULL terminator
+        // at address 63
+        char name_v1[GSD_NAME_SIZE];
+        strncpy(name_v1, name, GSD_NAME_SIZE - 1);
+        name_v1[GSD_NAME_SIZE - 1] = 0;
+        gsd_byte_buffer_append(&handle->frame_names.data, name_v1, GSD_NAME_SIZE);
+        handle->frame_names.n_names++;
+
+        // update the name/id mapping with the truncated name
+        int retval = gsd_name_id_map_insert(&handle->name_map, name_v1, *id);
+        if (retval != GSD_SUCCESS)
+        {
+            return retval;
+        }
+    }
+    else
+    {
+        gsd_byte_buffer_append(&handle->frame_names.data, name, strlen(name) + 1);
+        handle->frame_names.n_names++;
+
+        // update the name/id mapping
+        int retval = gsd_name_id_map_insert(&handle->name_map, name, *id);
+        if (retval != GSD_SUCCESS)
+        {
+            return retval;
+        }
+    }
+
+    return GSD_SUCCESS;
+}
+
+/** @internal
+    @brief Truncate the file and write a new gsd header.
+
+    @param fd file descriptor to initialize
+    @param application Generating application name (truncated to 63 chars)
+    @param schema Schema name for data to be written in this GSD file (truncated to 63 chars)
+    @param schema_version Version of the scheme data to be written (make with gsd_make_version())
+*/
+static int
+gsd_initialize_file(int fd, const char* application, const char* schema, uint32_t schema_version)
+{
+    // check if the file was created
+    if (fd == -1)
+    {
+        return GSD_ERROR_IO;
+    }
+
+    int retval = ftruncate(fd, 0);
+    if (retval != 0)
+    {
+        return GSD_ERROR_IO;
+    }
+
+    // populate header fields
+    struct gsd_header header;
+    gsd_util_zero_memory(&header, sizeof(header));
+
+    header.magic = GSD_MAGIC_ID;
+    header.gsd_version = gsd_make_version(2, 0);
```

Comment:
Should we define an enum at the top of this file defining the current version, so that we don't have to update this in multiple places?

Reply:
I added it, but many of those locations require an explicit 2.0 check. If we ever add 3.0, the conditions need to handle 1.0, 2.0, and 3.0 cases explicitly and not based on whatever the current spec is.

Reply:
Yes, I realize that some of them are checks specific to a given API. Having the enum does make it easier to not make mistakes when updating, though, since it saves some effort in looking at which instances are specific version checks and which are the current version.

Reply:
Agreed.